subsys/net/lib/coap/coap_transport_socket.c

```diff
@@ -231,6 +231,25 @@ static coap_transport_handle_t socket_create_and_bind(u32_t index,
 	socklen_t address_len = address_length_get(local->addr);
 
 	if (socket_fd != -1) {
+		if (local->interface) {
+#ifdef CONFIG_BSD_LIBRARY
+			
+#include "nrf_socket.h"
+
+			struct nrf_ifreq ifr;
+			memset(&ifr, 0, sizeof(ifr));
+			memcpy(ifr.ifr_name, local->interface, strlen(local->interface));
+
+			int err = nrf_setsockopt(socket_fd, NRF_SOL_SOCKET, NRF_SO_BINDTODEVICE, 
```

Comment:
This is now possible to use `setsockopt` as suggested by @stig-bjorlykke, please apply this change.

Reply:
The PR #448 is incomplete. Cannot use the data type struct ifreq needed for the socket option!
Are you suggesting I fix this first and then advance my PR?

Reply:
I'm a little bit lost here - what's exactly causing the problem? If there's a problem with `setsockopt`/`getsockopt` wrappers we should fix them. Not necessirally int this PR, but at least I'd like to understand what needs to be fixed.

Reply:
I still have to continue using nrf_socket.h as the data type needed to pass the parameters for the option SO_BINDTODEVICE are unavailable. Ideally the data type should have been introduced when SO_BINDTODEVICE was introduced. But it is not.

So, now I pushed code where the options from the socket.h but the data type is from nrf_socket.h. This is not neat, and is confusing.

But this is the quickest way to conform to your request.

Reply:
> But this is the quickest way to conform to your request.

I'm pretty sure I was not the original author of this request, but I back it, because it's the right path to follow IMO.

> So, now I pushed code where the options from the socket.h but the data type is from nrf_socket.h. This is not neat, and is confusing.

Yes it is. And I do not understand why do we need to mix these two namesapces here. `SO_BINDTODEVICE` is a well known option with a well know definition in man pages, which clearly states that:
```
The passed option is a variable-length null-terminated interface name string
```
So the type of the option is clearly defined - it's a C string. I don't see a reason why `socket.h` should cover it with a typedef.

If, for whatever reason, nrf_sockets accepts this option in a different format, then we should translate at the bsdlib shim layer. I tried really hard to find `struct nrf_ifreq` definition to verify if that's needed or not, but I failed. That's probably the reason why the CI is still failing.

Reply:
Two things to consider here:
- what does SO_BINDTODEVICE generally accept as a parameter data type? struct ifreq.
- can we change our implementation to accept a null terminated string? Of course we can!

But just because we can, should we? The struct ifreq is more flexible and in context of non APN based applications more extendible, we could bind to hardware identifiers, address or interface name etc. See http://man7.org/linux/man-pages/man7/netdevice.7.html.

Plus, we do not violate any standard interfaces if we use struct ifreq.

I am not sure I understand your comment on finding struct nrf_ifreq. This structure has been introduced with PDN interface management. The bsd lib PR pending should have this definition. 
See: https://github.com/NordicPlayground/nrfxlib/pull/27/files

Reply:
I just have to clarify that I would like to use SO_BINDTODEVICE and not NRF_SO_BINDTODEVICE. However, in current state, this interface is very confusing for me.

Reply:
> what does SO_BINDTODEVICE generally accept as a parameter data type? struct ifreq.

Well that stands in contradiction to http://man7.org/linux/man-pages/man7/socket.7.html. `SO_BINDTODEVICE` description does not mention `struct ifreq` at all. And neither http://man7.org/linux/man-pages/man7/netdevice.7.html mentions `SO_BINDTODEVICE` or even `setsockopt` - only `ioctl`.

But if you state that using `struct ifreq` as an argument is more future proof, then sure we can use it. I can add definition to the `nrfconnect-zephyr`, there's currently an update pending.

> I am not sure I understand your comment on finding struct nrf_ifreq. This structure has been introduced with PDN interface management. The bsd lib PR pending should have this definition.
See: https://github.com/NordicPlayground/nrfxlib/pull/27/files

And that's exactly the reason I did not find it. We cannot merge a PR that has dependencies which are not yet present in the master of `nrfxlib`. This breaks CI for this PR and will break CI for any other PRs until nrfxlib is updated. We could try to extract the changes needed for this PR and merge them rightaway into `nrfxlib` to solve this.